### ISSUE TITLE
Assertion error (`assert op == "=="`) in symbolic_shapes.prettify_results for operator `<=`

### ISSUE THREAD

**Issue by stellaraccident** [pytorch/pytorch#121286]
### 🐛 Describe the bug

While iterating on getting torch.export to work on a model with dynamic shapes, I hit the assertion around line 1640 of `symbolic_shapes.py` for the operator `<=`. I don't have a good repro as this was in the middle of a debugging session and the issue came from Torch attempting to print an error report for a problem I subsequently fixed.

I made the following changes and it got me past the issue:

```
            for s in self._static_results.union(self._dynamic_results):
                t = transform(s)
                if t == s:
                    continue
                left, op, right = re.split(r"( == | <= | >= | < | > )", t)
                op = op.strip()
                if op == "==" and left == right:
                    continue
                if right.isdigit():
                    relation_with_digit(left, op, int(right))
                elif left.isdigit():
                    relation_with_digit(right, flip(op), int(left))
                else:
-->                 if op == "<=":
-->                     results[left]["le"] = sympy.sympify(right)
                    else:
                        assert op == "==", f"Unexpected prettify op: {op}"
                        results[left]["eq"] = sympy.sympify(right)
```

We should probably add branches for each of the operators matched in the split a few lines above instead of just asserting on anything but `==`.

### Versions

```
torch==2.3.0.dev20240304+cpu
```

**Comment by ezyang** [comment]
@avikchaudhuri

**Comment by avikchaudhuri** [comment]
@stellaraccident Can you please report what expression you saw on the right of `<=`? I'm not expecting anything other than a constant there, so looking to figure out what's going wrong upstream of this logic. Relatedly, do you also see a log warning `"Failed to reduce inequalities: ...` before this happens? 

I'm planning to clean this up and some things should never make its way to the pretty printer because we're trying to suggest fixes here and there's no good fix for complicated inequalities.

**Comment by stellaraccident** [comment]
Ugh, I knew I should have stopped to get a repro and now I'm having trouble reconstructing it (this was a relatively involved debugging session). Here is the final state of what I was working on for reference: https://github.com/nod-ai/SHARK-Turbine/blob/da53717dc5dbcf4fa1531f5bff43e6ffa22925a0/llm/turbine_llm/inference/export_paged_llm_v1.py

Going from memory, I think this was coming from a prior version of this sequence:

```
        block_dim = torch.export.Dim("block", max=2047 // 16)
        sl_dim = 16 * block_dim
        page_dim = torch.export.Dim("page")
        dynamic_shapes = {
            "tokens": {1: sl_dim},
            "seq_lens": {},
            "seq_block_ids": {1: block_dim},
            "cache_state": [{0: page_dim}],
        }
```

I believe this was happening in the prior state of:

```
        block_dim = torch.export.Dim("block", max=2047 // 16)
        sl_dim = torch.export.Dim("sl", max=2047)
```

And this prettify_results function was being called as part of the error reporting for when a suggested change to the constraints is printed. Making the reported update to the code caused the error reporter to properly suggest the dependent dim fix:

```
        block_dim = torch.export.Dim("block", max=2047 // 16)
        sl_dim = 16 * block_dim
```

Without the patch, the error reporter was hitting that assertion. With the fix, I don't think the problematic constraint is printed and it doesn't go down this path at all. Iirc, there were definitely expressions involved, not just numbers.

Sorry, I assumed this would be obvious and should have gotten a better repro.

**Comment by stellaraccident** [comment]
Just tried a few things and can't get it walked back to where the assert was happening. Maybe we just extend the assert to print some more diagnostics on failure for now?

**Comment by ezyang** [comment]
I have a repro for this. Checkout https://github.com/ezyang/pytorch/tree/assert-op-fail and run `python test/export/test_export.py -k test_crop_like`

dynamic logs is https://gist.github.com/ezyang/9db41bfbbad0f36cd7f4f7f0aff98375

**Comment by IvanKobzarev** [comment]
Hi @avikchaudhuri , doing issue scrapping, is this issue is still relevant and you are working on it?

**Comment by avikchaudhuri** [comment]
> Hi @avikchaudhuri , doing issue scrapping, is this issue is still relevant and you are working on it?

Still relevant I think (`test_crop_like` still has a xfail), not working on it but thanks for reminding.
